jsmn.h

```diff
@@ -126,64 +166,244 @@ static void jsmn_fill_token(jsmntok_t *token, const jsmntype_t type,
+            numstate == JSMN_NUM_INT_HAVE_SIGN) {
+          numstate = JSMN_NUM_INT_ZERO;
+        } else if (numstate == JSMN_NUM_INT) {
+          numstate = JSMN_NUM_INT;
+        } else if (numstate == JSMN_NUM_FRAC ||
+                   numstate == JSMN_NUM_FRAC_START) {
+          numstate = JSMN_NUM_FRAC;
+        } else if (numstate == JSMN_NUM_EXP ||
+                   numstate == JSMN_NUM_EXP_START ||
+                   numstate == JSMN_NUM_EXP_HAVE_SIGN) {
+          numstate = JSMN_NUM_EXP;
+        } else {
+          parser->pos = start;
+          return JSMN_ERROR_INVAL;
+        }
+        break;
+      case '1':
+      case '2':
+      case '3':
+      case '4':
+      case '5':
+      case '6':
+      case '7':
+      case '8':
+      case '9':
+        if (numstate == JSMN_NUM_INT ||
+            numstate == JSMN_NUM_INT_START ||
+            numstate == JSMN_NUM_INT_HAVE_SIGN) {
+          numstate = JSMN_NUM_INT;
+        } else if (numstate == JSMN_NUM_FRAC ||
+                   numstate == JSMN_NUM_FRAC_START) {
+          numstate = JSMN_NUM_FRAC;
+        } else if (numstate == JSMN_NUM_EXP ||
+                   numstate == JSMN_NUM_EXP_START ||
+                   numstate == JSMN_NUM_EXP_HAVE_SIGN) {
+          numstate = JSMN_NUM_EXP;
+        } else {
+          parser->pos = start;
+          return JSMN_ERROR_INVAL;
+        }
+        break;
+      case '+':
+        if (numstate == JSMN_NUM_EXP_START) {
+          numstate++;
+        } else {
+          parser->pos = start;
+          return JSMN_ERROR_INVAL;
+        }
+        break;
+      case 'e':
+      case 'E':
+        if (numstate == JSMN_NUM_INT_ZERO ||
+            numstate == JSMN_NUM_INT ||
+            numstate == JSMN_NUM_FRAC) {
+          numstate = JSMN_NUM_EXP_START;
+        } else {
+          parser->pos = start;
+          return JSMN_ERROR_INVAL;
+        }
+        break;
+      case '.':
+        if (numstate == JSMN_NUM_INT_ZERO ||
+            numstate == JSMN_NUM_INT) {
+          numstate = JSMN_NUM_FRAC_START;
+        } else {
+          parser->pos = start;
+          return JSMN_ERROR_INVAL;
+        }
+        break;
+      default:
+        goto check_number_complete;
+      }
+    }
+check_number_complete:
+    if (numstate != JSMN_NUM_INT &&
+        numstate != JSMN_NUM_INT_ZERO &&
+        numstate != JSMN_NUM_FRAC &&
+        numstate != JSMN_NUM_EXP) {
+      if (parser->pos >= len || js[parser->pos] == '\0') {
+        parser->pos = start;
+        return JSMN_ERROR_PART;
+      } else {
+        parser->pos = start;
+        return JSMN_ERROR_INVAL;
+      }
+    }
+  }
+
+  /* Verify that what comes after the primitive is a non-primitive character */
+  if (parser->pos < len) {
     switch (js[parser->pos]) {
-#ifndef JSMN_STRICT
-    /* In strict mode primitive must be followed by "," or "}" or "]" */
+    case '\t':
+    case '\r':
+    case '\n':
+    case ' ':
+    case ',':
     case ':':
```

Comment:
According to the rfc, I don't believe that a primitive can be followed by a `':'`, `'{'`, or `'['` and can only be followed by `'\0'` when it is the lone element in the json string.

Reply:
This switch's purpose is simply to ensure there aren't extra characters in the primitive, like `1.234true`. As long as the character after is not allowed in a primitive, the check is satisfied. Proper validation of the next character happens elsewhere (in the main parsing loop).

This is important because jsmn will parse multiple "JSON texts" (in RFC 8259 terminology) in one string, if there are more than one present. For example, this is perfectly valid input for jsmn and will simply parse to two separate arrays:

    ["a", "b", "c"] [1, 2, 3]

Since this pull request adds the ability for strings and primitives to be complete JSON texts on their own, this check ensures that `1.234true` will be rejected rather than parsed as two separate primitives. (Strictly speaking, you could say they *should* be parsed as two primitives, since consecutive objects and arrays need not be separated by whitespace to be parsed by jsmn. But, consider a case like `1.234.5`: this would become ambiguous under those rules. It could be `1.2 34.5` or `1.23 4.5`. For that matter, even `1.234true` could be `1.234 true`, `1.23 4 true`, `1.2 34 true`, etc. Requiring at least one non-primitive character as separation resolves this ambiguity.)

Parsing multiple JSON texts is also the reason we might see a `{` or `[` following a primitive. `:` cannot follow a primitive in strict RFC 8259 compliance, but it can in non-strict mode. In another pull request (#195), I add the ability to select specific non-strict features, so it is possible to have strict primitive validation but allow primitives as object keys.

Reply:
These are all valid input strings in strict mode:
- `true["a", "b", "c"]` (Parses as a primitive followed by an array. `[` follows a primitive.)
- `true{"a": 1}` (Parses as a primitive followed by an object. `{` follows a primitive.)
- `true` (Parses as an unenclosed primitive. `\0` follows a primitive.)

Then, regarding the colon:

`:` can never follow a primitive in strict mode, but I put it there anticipating enhancements like PR #195, which is built on top of this one. I encourage anyone reading to take a look at that PR for the full details, but to summarize the relevant part: It takes each use of `#if[n]def JSMN_STRICT` and replaces it with `#if[n]def JSMN_PERMISSIVE_PRIMITIVES`, `JSMN_PERMISSIVE_STRINGS`, or `JSMN_PRIMITIVE_KEYS`, allowing specific non-strict features to be selected. Notably, this means if `JSMN_PRIMITIVE_KEYS` is defined but `JSMN_PERMISSIVE_PRIMITIVES` is not, that `:` in the switch statement becomes important.